### PR TITLE
Early Vote Site Tool Epic PR

### PR DESCRIPTION
**Pull request by tank157**

### CODE REVIEW

**Comment by tank157** [comment]
Basically this now covers most/all the work for [this epic](https://www.pivotaltracker.com/epic/show/3714455). The cards that are open have been annotated to point out where in the code we implement these features.

File: dashboard/util.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+          var params = paramFns[i](req);
+          if (params) {
+            for (var j = 0; j < params.length; j++) {
+              ret.push(params[j]);
+            }
+          }
+        }
+      return ret;
+    }
+  },
```
Comment (tie-rack) [inline dashboard/util.js]: These extractors are so nearly `map` and `mapcat` they _could_ be:

```js
queryParamExtractor: params => req => params.map(param => decodeURIComponent(req.query[param]))
pathParamExtractor: params => req => params.map(param => decodeURIComponent(req.params[param]))
bodyParamExtractor: params => req => params.map(param => req.body[param])
compoundParamExtractor: paramFns => req => paramFns.reduce((allParams, paramFn) => allParams.concat(paramFn(req)), [])
```

Maybe you don't have to go that far, but using `map` in place of all the `for (var i = 0; i < params.length; i++)` bits is worth not reinventing `Array.prototype` functions.

File: early-vote/assignments.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                                           util.pathParamExtractor(['assignmentid']));
+
+//create assignment
+var createSql = "INSERT into assignments(id, early_vote_site_id, schedule_id) " +
+  "values ($1, $2, $3);"
+
+var createParamsFn =
+  util.compoundParamExtractor([util.uuidGenerator(),
+                               util.pathParamExtractor(['earlyvotesiteid']),
+                               util.bodyParamExtractor(['schedule_id'])]);
```
Comment (tie-rack) [inline early-vote/assignments.js]: So _functional_.

File: early-vote/file-generation.js [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+/**
+ * Given an array of schedules, creates a map where the schedules are
+ * keyed by id for easy lookup.
+ */
+var schedulesById = function(schedules) {
+  var byId = {};
+  schedules.forEach(function(schedule) {
+    byId[schedule["id"]] = schedule;
+  });
+  return byId;
```
Comment (tie-rack) [inline early-vote/file-generation.js]: This function and the next two have that classic `reduce` look to them.

**Comment by tank157** [comment]
@tie-rack Ok, I've taken your suggestions and run with them. Had some difficulty in some of the reduces, had to revert to `hash[key] = newval` syntax that I wasn't as thrilled with but trying to use the new `Map` type didn't work out, nor did other things I tried. Anyway, the arrow functions are neat, I'd only been vaguely aware of them, so getting pointed to actual documentation on them helped a lot!

File: public/assets/css/re-frame.css [hunk 4]
```diff
@@ -1,4 +1,4 @@
-/* variables-and-mixins.scss */
+breadcrumb/* variables-and-mixins.scss */
```
Comment (tie-rack) [inline public/assets/css/re-frame.css]: This `breadcrumb` looks out of place

**Pull request merged**